### ISSUE TITLE
Consider dropping support for Visual C++ versions earlier than VC++ 2019, x86 builds, 32-bit ARM, and setting minimum expectations on SIMD support

### ISSUE THREAD

**Issue by spectecjr** [openmm/openmm#2656]
HI there! First time caller... you probably know the rest :)

I've been poking around and looking at the code, and would like to make some recommendations. Obviously I'm new to OpenMM, so there may well be great reasons for these suggestions not to happen that I'm just not aware of. 

Just to make keeping track of builds easier, it might be worth dropping support for some categories of devices and environments, and setting some minimum expectations on SIMD feature support.

**Consider dropping support for versions of VC++ earlier than 2019**

Since VS2015, the Community edition of Visual C++ has incorporated the full optimizer of the Pro edition, and all other editions have moved to a subscription model, which means that there's no good reason to stay on earlier versions unless absolutely necessary. (Internally, Microsoft has shifted heavily towards focusing their efforts on the latest version, so all new feature work, feature improvements, bugfixes, and optimization improvements focus there).

As both VS2019 and Clang are effectively free for small users, and for larger groups the compiler has been available by subscription since 2015 (meaning that if developers have already made the leap, they can upgrade to the latest version for no extra cost), it's worth moving forward so that future development can take advantage of bugfixes and performance improvements in C++17 and up. 

**Consider dropping explicit 32-bit Intel/AMD support, possibly 32-bit ARM support**

x86 is effectively an obsolete platform at this point - it was introduced by AMD in 1999, and according to https://www.pcbenchmarks.net/os-marketshare.html , 32-bit installations of operating systems make up ~1% of all installations today. 

With the release of MacOS Catalina last year, 32-bit support is no longer provided by Apple in their operating systems releases. Beyond obsolesence, with the larger address space giving serious advantages when dealing with large datasets and working sets. x64 also has better performance when throwing exceptions due to its table-driven unwinding mechanism, and a more efficient ABI.

Google currently requires all apps that ship in the Google Play store to provide a 64-bit version (ARM8/AARCH64) of their app (the 32-bit version is optional). This introduces a number of features by default, including Neon support. 95% of Android devices support 64-bit Android OSes (Android v5+ - see https://www.appbrain.com/stats/top-android-sdk-versions - stats are generated from roughly 100MM unique users).

**Consider supporting AVX as your minimum SIMD platform on Intel/AMD, ideally incorporate support for AVX2**

Both Clang and Visual C++ 2019 are able to symbolically re-optimize code written using SIMD intrinsics - often (and perhaps surprisingly) better than the original programmer. This means that SIMD platforms without explicit support can often be auto-optimized to use SIMD instructions based on the compiler's ISA flag settings. 

Therefore, it's possible to target a minimum SSE support level for optimization, while leaving the straight C++ version of the routine in place, both as a sanity check mechanism for testing (to ensure the results are the same as the unoptimized version), and as a way to support downlevel CPUs without necessarily requiring explicit programmer support.

In March 2018, Microsoft dropped support for CPUs without at least SSE2 feature support, primarily affecting Pentium III processor users. Windows 8 requires SSE2 support. It is reasonable to set SSE2 as a minbar.

For what it's worth, looking at the CMakeLists.txt files, the current minimum ISA support required is for AVX and/or SSE4.1.

**Probably should leave PowerPC alone**
Two of the top supercomputers in the world right now (according to the TOP500 list) run PowerPC, and they might need OpenMM, so we should probably leave that functionality alone for now - that said, Xbox 360 and PS3 are effectively obsolete, and any effort to continue supporting it **may** be unnecessary. 

I can't make a recommendation here though -  I'll just raise the question of if it makes sense to continue PowerPC support.

**Comment by peastman** [comment]
Thanks for the comments!  Could you elaborate on what specific changes you're suggesting?  Understand that we don't do anything in particular to "support" 32 bit x86, 32 bit ARM, or Visual Studio 2015.  It's just C++ code.  You can compile it in whatever mode you want.  All of our pre-built packages are for 64 bit architectures, but if someone wanted to compile for 32 bit mode, I don't know of any reason it wouldn't work.

**Comment by dwtowner** [comment]
FYI, it is my intention to add AVX2 and AVX-512 support as part of the work I'm doing on #2661. I'll keep SSE though still.

**Comment by spectecjr** [comment]
I may have only half thought this through, but here's my intent.

If  you can set a minimum bar for support, it reduces the size of the test (and support) matrix and allows people to focus on optimizations for the smaller set of platforms you do support.

For example, if you could say:

**Language support**
We support features found in C++ 17 and earlier.

**Building**
For Windows builds, we support 64-bit builds running on Windows 7 or higher, building with Visual C++ 2019 Community Edition.
For Linux builds, we support (version details), building with Clang v. xxxx, LLD v..
For MacOS builds we support (foo), building with (bar).
For Android builds we support 64-bit builds running on Android API Level 28 or later, with NDK 21B.

**CPU Hardware features**
x64 with AVX, SSE, SSE2, SSE4.1, with fast-math enabled.
64-bit ARM with ISA armv8-a, thumb-enabled, NEON SIMD instructions, SVE2, integer divide, floating point hardware, and the aarch64 ABI.  FP is set to fast mode. We optimize both NEON and SVE2 builds.

We support all other architectures by allowing Clang/LLVM to auto-vectorize the code; no specific optimizations are required for PowerPC moving forward, although existing ones are not deprecated. 

--------------
Basically, I'm trying to reduce the targetable surface area to the common cases. In retrospect this is probably unnecessary because as you say, you kind of get the other builds for free. It'd still be good to have an explicit policy spelled out as to what people need to support and what they don't. (Some of this was triggered by reading **dwtowner**'s attempted checkin).

**Comment by spectecjr** [comment]
One thing that may be a useful change is the codebase should probably migrate forwards to a later version of C++ so that it can drop its pthreads dependency and use std::thread.

**Comment by dwtowner** [comment]
std::thread is part of C++11, which is already used anyway.

**Comment by dwtowner** [comment]
> FYI, it is my intention to add AVX2 and AVX-512 support as part of the work I'm doing on #2661. I'll keep SSE though still.

AVX2 now supported as part of #2724. AVX-512 working locally but needs a bit more tidying before I do a PR.
